iml-services/iml-devices/src/breadth_first_parent_iterator.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use iml_wire_types::db::{Device, DeviceId};
+use std::collections::{BTreeMap, BTreeSet};
+
+pub struct BreadthFirstParentIterator<'a, 'b> {
+    devices: &'a BTreeMap<DeviceId, Device>,
+    parents: BTreeSet<DeviceId>,
+    next_parents: BTreeSet<DeviceId>,
+    _marker: &'b std::marker::PhantomData<()>,
+}
+
+impl<'a, 'b> BreadthFirstParentIterator<'a, 'b> {
+    pub fn new(devices: &'a BTreeMap<DeviceId, Device>, device_id: &'b DeviceId) -> Self {
+        tracing::trace!("Getting {:?} from devices", device_id);
+        // TODO: This is dangerous
```

Comment:
??

Reply:
I've documented caller responsibility so this doesn't panic.

In practice this is never the case since we always start with device that's already in `devices`, and tree as given by `device-scanner` is always consistent and contains all parents.

Reply:
I believe there is some logical issues here.

First of all it appears that you use the iterator only once (in `are_all_parents_available`) and thus I doubt it needs to be an iterator at all.

Then, I don't think anything can justify a possible crash at such a low level.

So the question is, if a crash happens, what shall we do? Shouldn't we instead catch and report  invalid data somehow?

I'd make `are_all_parents_available` return a `Result<bool, _>`. If you want a crash, you will be able to `unwrap()` :)

Reply:
This needs a separate code module to be able to test it and reason about it independently. Reason `are_all_parents_available` is so simple is because iterator is separate.

Guarding for `device-scanner` issues is not the purpose of this code. If `device-scanner` starts to send incorrect data we'll see it just fine with `iml-devices` panicking. It's probably even better than trying to recover from that, because `device-scanner` is a component pretty much everything depends on, and if it doesn't work correctly, it doesn't make much sense to try.

Next, issue is not `are_all_parents_available`. To catch an error we'll need the iterator to return a `Result` and it will make a mess of clean `i.all(...)` that we have now, for nothing.